firebase_admin/db.py

```diff
@@ -768,46 +771,110 @@ class _DatabaseService(object):
+            else:
+                credential = self._credential.get_credential()
+            client = _Client(credential, base_url, self._timeout, params)
+            self._clients[client_cache_key] = client
+        return self._clients[client_cache_key]
 
     @classmethod
-    def _validate_url(cls, url):
-        """Parses and validates a given database URL."""
+    def _parse_db_url(cls, url, emulator_host=None):
+        """Parses a database URL into (base_url, query_params, use_fake_creds).
+
+        The input can be either a production URL (https://foo-bar.firebaseio.com/)
+        or an Emulator URL (http://localhost:8080/?ns=foo-bar). The resulting
+        base_url never includes query params. Any required query parameters will
+        be returned separately as a map (e.g. `{"ns": "foo-bar"}`).
+
+        If url is a production URL and emulator_host is specified, the result
+        base URL will use emulator_host, with a ns query parameter indicating
+        the namespace, parsed from the production URL. emulator_host is ignored
+        if url is already an emulator URL. In either case, use_fake_creds will
+        be set to True.
+        """
         if not url or not isinstance(url, six.string_types):
             raise ValueError(
                 'Invalid database URL: "{0}". Database URL must be a non-empty '
                 'URL string.'.format(url))
-        parsed = urllib.parse.urlparse(url)
-        if parsed.scheme != 'https':
-            raise ValueError(
-                'Invalid database URL: "{0}". Database URL must be an HTTPS URL.'.format(url))
-        elif not parsed.netloc.endswith('.firebaseio.com'):
+        parsed_url = urllib.parse.urlparse(url)
+        use_fake_creds = False
+        if parsed_url.netloc.endswith('.firebaseio.com'):
+            if parsed_url.scheme != 'https':
+                raise ValueError(
+                    'Invalid database URL: "{0}". Database URL must be an HTTPS URL.'.format(url))
+            # pylint: disable=invalid-name
+            base_url, namespace = cls._parse_production_url(parsed_url)
+            if emulator_host:
+                base_url = 'http://{0}'.format(emulator_host)
+                use_fake_creds = True
+        else:
+            use_fake_creds = True
+            base_url, namespace = cls._parse_emulator_url(parsed_url)
+
+        if not base_url or not namespace:
             raise ValueError(
                 'Invalid database URL: "{0}". Database URL must be a valid URL to a '
                 'Firebase Realtime Database instance.'.format(url))
-        return 'https://{0}'.format(parsed.netloc)
+        if base_url == 'https://{0}.firebaseio.com'.format(namespace):
+            # namespace can be inferred from the base_url. No need for query params.
+            return base_url, {}, use_fake_creds
+        else:
+            # ?ns=foo is needed.
+            return base_url, {'ns': namespace}, use_fake_creds
```

Comment:
I tried that but I found the return value `ns` really confusing, because it's not always the namespace name. It may be None if not needed, etc. and cannot be used for anything other than param. If we were to go with simplifying `params`, then I would actually move the param logic entirely out of the parsing method, and have something like:

```python
base_url, namespace = _parse_db_url()  # namespace will never be None

if base_url == 'https://{0}.firebaseio.com'.format(namespace):
    use_fake_creds = False
    params = {}
else:
    use_fake_creds = True
    params = {'ns': ns}
```

And we can further simply the parsing method. WDYT?

Reply:
I kind of like that. You can also get rid of the use_fake_creds by doing this:

```
base_url, namespace = _parse_db_url()  # namespace will never be None

if base_url == 'https://{0}.firebaseio.com'.format(namespace):
    credentials = self._credentials.get_credentials()
    params = {}
else:
    credentials = _EmulatorAdminCredentials()
    params = {'ns': namespace}
```

Reply:
Yay, done!